Get rid of Timeline runtime repair

cstamas opened the pull request:
Repair should be tried only on "boot". During runtime, if corruption is detected, the Indexer should be simply shot down, marked as "broken" and no further operation tried against it.

On next boot, it will be repaired (or tried at least), but no data loss occurs, as the persistor will still write the records.

cstamas committed:
Getting rid of runtime repair.

Repair is tried only on boot, and during runtime,
if indexer problem is detected, it is simply made
inactive and is shut down.

cstamas committed:
markIndexerDead should handle IOEx

By logging, and not disturbing caller with it, as caller
called us already coz it had one.

ifedorenko commented on nexus/plugins/nexus-timeline-plugin/src/main/java/org/sonatype/timeline/internal/DefaultTimeline.java:
```diff
@@ -247,20 +274,18 @@ E doIt()
     {
         if ( timelineLock.readLock().tryLock() )
         {
             if ( started && !indexerIsDead )
```
Lock is not unlocked if  "started && !indexerIsDead" evaluates to false.

ifedorenko commented:
-1

doShared is both incorrect (see the comment above) and is generally too complicated. IOException during work.doIt() most likely means something really bad, like filesystem corruption or too many open files. Get exclusive lock and do markIndexerDead does not add much, something like this should be sufficient

```
protected <E> E doShared( final Work<E> work )
{
    if ( timelineLock.readLock().tryLock() )
    {
        try
        {
            if ( started && !indexerIsDead )
            {
                try
                {
                    return work.doIt();
                }
                catch ( IOException e )
                {
                    indexerIsDead = true;
                    getLogger().warn( "Timeline index got corrupted and is disabled. Repair will be tried on next boot.",
                                      e );
                }
            }
        }
        finally
        {
            timelineLock.readLock().unlock();
        }
    }
    return null;
}
```

This will also allow simpler and cleaner start() code and get rid of doExclusively.

Nitpicking, remove unused private TimelineConfiguration configuration member.

jdillon commented:
How about this:

```
rm -rf nexus/plugins/nexus-timeline-plugin
```

cstamas committed:
Get rid of timeline repair.

Applied Igor's comments.

cstamas commented:
Applied Igor's comments. Jason's comment not taken seriously yet.

ifedorenko commented:
Nitpicking. "if !started" checks everywhere are redundant, the check in doShared is enough. Otherwise +1, pending outcome of the discussion if nexus should or should not start if timeline cannot be opened.

cstamas committed:
Timeline index repair failure is not propagated.

Timeline handles indexer with special care: it will try
to repair it if it's start failed. This change now
will not propagate anymore the exception thrown during
repair (already in exception handler!), but will mark the
indexer as "dead", log the fact, and continue the usual
fluff.

On next reboot, another repair will be tried.

adreghiciu commented:
+1

cstamas commented:
CI liked this too (linux slave is blue, did not wait for win slave):
https://builds.sonatype.org/view/nexus-features/job/nexus-oss-its-feature/521/
